packages/dds/tree/docs/wip/Container and Application Schema.md

```diff
@@ -0,0 +1,194 @@
+But I think we can pick a good subset for the now, and leave open the option of extending it in the future.
+
+I think we want to at least:
+
+-   Make adoption of different constraint/schema systems optional and incremental: all of these different options should be opt in, and easy to add to and existing application if desired.
+-   Provide a schema-on-read system (ex: schematize) that can give an application a nice interface to the data, including validation and error handling
+    Anytime the application schema and container schema are different, this can allow the app to function as well as possible (detect and handle out of schema data, both on read and on edit, as localized as permitted by the provided error handlers).
+-   Provide at least one way to customize the container schema (schema-on-write), which can at least do type based constraints.
+-   Have at least one good design-pattern apps can use to handle schema migrations for each way they have to enforce/specify schema (make sure this supports a roll out process were clients have mixed code versions, and old documents will be supportable forever)
+
+# Expected usage patterns
+
+I suspect different users will mostly fall into 2 groups:
+
+-   pure schema-on-read: no customization of container schema.
+-   almost pure schema-on-write: only using schema-on-read to assist with schema migration so their application schema doesn't have to deal with old formats (which unavoidably pile up in the container schema)
+    They may additionally use schema-on-read to enforce/fix some invariants the container schema system isn't expressive enough to declare (referential integrity for graph like references, numeric ranges, and any other appellation level invariants that are desired)
+
+# Example/Proposed Design
+
+This design shows an easy to implement MVP, and is not intended to have document or API compatibility with what we do longer term.
+It minimally implements the requirements from "What options to support" while being extensible in the future.
+It can be subseted for use in earlier milestones, and could be extended for later ones, see "Schedule" below.
+
+This design is intentionally minimal while still being useful.
+Features not needed for minimal strongly typed application use or the JSON and XML domain schema are not included.
+
+## Schema Representation
+
+Include a simple non-contextual schema system, usable as a MVP for both `container schema` and `application schema`.
+
+We can start with:
+
+```typescript
+enum Multiplicity {
+    Value,
+    Optional,
+    Sequence,
+}
+
+enum Value {
+    Nothing,
+    Number,
+    String,
+    Boolean,
+    Serializable,
+}
+
+interface FieldContent {
+    multiplicity: Multiplicity;
+    types?: Type[];
+}
+
+interface Field {
+    content: FieldContent;
+    name: string;
+}
+
+interface Type {
+    name: string;
+    fields: Field[];
+    extraFields: FieldContent;
```

Comment:
The purpose of the schema system is to restrict what data is allowed. While some applications might want to allow all possible values for unrecognized fields, doing so should be optional.

There are use-cases for preventing such flexible augmentations:
- It's much easier to add fields to nodes you know don't already have those fields.
- Some users may want to ensure specific nodes don't have extra data (ex: number terminals for some users)
- Its easier to properly maintain data the app does not understand if you limit it to occurring in a few explicit locations instead of literally anywhere.

There are some use-cases for allowing it anywhere as well.

What we want the defaults to be in the schema language, and what we want to encourage applications to do pattern wise is a separate issue.

At this abstraction level, there is no distinction between augmentations and other use cases for fields. I consider augmentations to be an optional design pattern which might be implemented using extraFields but could also be implemented as a field containing a list of augmentations, or an augmentations child node in a value trait.

Reply:
The local/global split seems to have helped, in that you can still constraint the 'V' of a Map<K, V> using 'localExtraFields' while allowing augmentation via global fields.